File: tests/xcalablemp/not-implemented-yet/global-view/tasks/C/tasks.c
/*testp065.c*/
/*tasks指示文およびtask指示文のテスト*/
#include<xmp.h>
#include<mpi.h>
#include<stdio.h>
#include<stdlib.h>      
static const int N=1000;
#pragma xmp nodes p(*)
#pragma xmp template t(0:N-1,0:N-1,0:N-1)
#pragma xmp distribute t(*,*,block) onto p
int a[N],s,lb,ub,procs ,w,ans;
int i;
char *result;
#pragma xmp align a[i] with t(*,*,i)
#pragma xmp shadow a[1]
int main(void){

   s=0;
#pragma xmp loop on t(:,:,i)
   for(i=0;i<N;i++){
      a[i]=i;
   }

   procs = xmp_num_nodes();

   if(N%procs == 0){
      w = N/procs;
   }else{
      w = N/procs+1;
   }

   lb = (xmp_node_num()-1)*w;
   ub = xmp_node_num()*w;

#pragma xmp tasks
   {
#pragma xmp task on p(1)
      {
         for(i=lb;i<ub+1;i++){
            s = s+a[i];
         }
      }
#pragma xmp task on p(2)
      {
         if(procs > 1){
            for(i=lb;i<ub+1;i++){
               s = s+a[i];
            }
         }
      }
#pragma xmp task on p(4)
      {
         if(procs > 3){
            for(i=lb;i<ub+1;i++){
               s = s+a[i];
            }
         }
      }
   }
#pragma xmp reduction (+:s)
   if(procs < 2){
      ans = w*(w+1)/2;
   }else if(procs < 4){
      ans = w*(w*2+1);
   }else{
      ans = w*(w*2+1) + w*2*(w*4+1) - w*3*(w*3+1)/2;
   }

   if(ans == s){
      result = "OK";
   }else{
      result = "NG";
   }

   printf("%d %s %s\n",xmp_node_num(),"testp065.c",result);
   return 0;
}    

         
      
   

File: tests/XMP/NOT-IMPLEMENTED/38/tasks.f90
! testp065.f
! tasks指示文およびtask指示文のテスト

      program main
      include 'xmp_lib.h'
      integer,parameter:: N=1000
!$xmp nodes p(*)
!$xmp template t(N,N,N)
!$xmp distribute t(*,*,block) onto p
      integer a(N), s, lb, ub, procs, w, ans
      character(len=2) result
!$xmp align a(i) with t(*,*,i)
!$xmp shadow a(1)

      s = 0
!$xmp loop on t(:,:,i)
      do i=1, N
         a(i) = i
      enddo

      procs = xmp_num_nodes()

      if(mod(N,procs) .eq. 0) then
         w = N/procs
      else
         w = N/procs+1
      endif

      lb = (xmp_node_num()-1)*w+1
      ub = xmp_node_num()*w
      
!$xmp tasks
!$xmp task on p(1)
      do i=lb, ub
         s = s+a(i)
      enddo
!$xmp end task
!$xmp task on p(2)
      if(procs .gt. 1) then
         do i=lb, ub
            s = s+a(i)
         enddo
      endif
!$xmp end task
!$xmp task on p(4)
      if(procs .gt. 3) then
         do i=lb, ub
            s = s+a(i)
         enddo
      endif
!$xmp end task
!$xmp end tasks

!$xmp reduction(+:s)

      if(procs .lt. 2) then
         ans = w*(w+1)/2
      else if(procs .lt.4) then
         ans = w*(w*2+1)
      else
         ans = w*(w*2+1) + w*2*(w*4+1) - w*3*(w*3+1)/2
      endif
      
      if(ans .eq. s) then
         result = 'OK'
      else
         result = 'NG'
      endif

      print *, xmp_node_num(), 'testp065.f ', result

      end
